README.md
# NRAP-Open-IAM
Clone of the original NRAP-Open-IAM code stored in GitLab


CHANGELOG.md
# Changelog
All notable changes to this project are documented in this file.

**Note (for all current and future maintainers of the code):** Whenever the version
of the tool is updated and changes are recorded in the current changelog,
please perform the following actions:
  - update version of the tool in the following files:
    - source/openiam/\__init\__.py
    - documentation/user/source/conf.py
    - documentation/developer/source/conf.py
    - documentation/qaqc/source/conf.py
    - documentation/website/source/conf.py
    - source/GUI/Dashboard.py
    - replace image of front page of GUI with the one stating new version
    of the tool in the documentation.
  - recompile user and developer guides, and QAQC documents;
  - add tag to the master branch corresponding to the current change log record;
  - replace old version of the user and developer guides with the new recompiled
  files in the current folder.
  - do not forget to commit the changes.


## [a2.7.2 - 2023-08-25]
### Added
  - CFI: Added new section for workflows; examples illustrating AoR workflow
  setup
  - Auxiliary component: Well Depth Risk Configurer
  - Jupyter notebooks illustrating setup and run of control file examples
  including workflows, and TTFD and AoR types of analysis and visualization
  - Output of the current NRAP-Open-IAM version and runtime information
  to a text file for control and GUI files.
### Updated
  - Auxiliary Location Generator component: generation of z-coordinates,
  in addition to x- and y-coordinates
  - Logging handlers bug fix
  - Moved CFI related files to a separate folder
  - Documentation updates for webpage

## [a2.7.1 - 2023-06-30]
### Added
  - Hydrocarbon Leakage component: class and CFI
  - Cemented Wellbore (WR) component: GUI and an example illustrating setup
  - Fault Leakage component: CFI and GUI; CFI and GUI examples illustrating setup
  - CFI: Stratigraphy column type of plot; examples illlustrating setup
  - CFI: Gridded radial metric type of plot for Generic Aquifer component;
  examples illustrating setup
  - CFI: Gridded metric type of plot for Seal Horizon and Fault Flow components;
  examples illustrating setup
  - Lookup Table Reservoir component: option for the use of lookup tables
  in *.h5 format; related tests to the test suite
  - Plume Stability component: option for the component metrics to be
  calculated for 3d data; option for the use of lookup tables in *.h5
  format; related tests to the test suite
  - Auxiliary functionality: well data classifier, leak scorer from well integrity
  study
  - Auxiliary components: Parameter Setup, Pressure Based Risk Configurer,
  Data Based Risk Configurer, Monitoring Scheduler, Monitoring Tool
  - Functionality to save gridded observations in formats different from *.npy
  - Documentation:
   - NRAP-Open-IAM workflows illustration figures
   - Section on conceptual model overview
### Updated
  - CFI and GUI for Open Wellbore component: option to consider critical
  pressure for leakage estimates
  - Plotting functionality related to sensitivity analysis in CFI and GUI;
  new CFI examples illustrating setup
  - Stratigraphy, time series, Total time to the first detection (TTFD), Area of Review
  workflows plots
  - Documentation:
    - Description of existing and added visualization capabilities

## [a2.7.0 - 2023-02-15]
### Added
  - Generic Reservoir component: class, CFI and GUI; script and CFI examples illustrating
  setup
  - Theis Reservoir component: class and CFI
  - Cemented Wellbore (WR) component: class and CFI (updated version of the previous
  cemented wellbore component); script and CFI examples illustrating setup
  - Fault Leakage component: class
  - Script example illustrating coupling of NRAP-Open-IAM and SOSAT tool
  - Additional options for setup of time series and AoR plots in CFI
  - Total time to the first detection (TTFD) workflow plots in CFI; examples
  illustrating setup
  - CFI: Stratigraphy setup through input file and its visualization through
  new stratigraphy plot
  - Script and CFI examples illustrating new visualization capabilities
  - GUI: options to save simulation results files for individual or combined outputs
### Updated
  - Setup of AmtPlumeSingle and AtmPlumeEnsemble plots
  - Documentation:
    - Description of recently added components
    - Description of new CFI examples
    - Description of existing and added visualization capabilities

## [a2.6.1 - 2022-09-30]
### Added
  - GUI: time points setup through manual and file inputs
  - CFI: new options for setting up locations for all placeable components
  (reservoir, leakage pathways, impact)
  - Area of Review (AoR) calculation script and CFI examples; AoR plot option
  - CFI: options to save simulation results files for individual or combined outputs
### Updated
  - Seal Horizon component class: added option to sample thickness and permeability
  parameters inside component; for CFI added option of different controls inherited
  from standalone Seal Flux code
  - Fault Flow component class: added option to sample strike, dip and aperture
  parameters; for CFI added option of different controls inherited
  from standalone FaultFlo code
  - Generic Aquifer component class: fixed issue with results for zero
  CO2 leaked mass
  - Restructured code related to CFI: the methods are split among several files
  according to the functionality they cover

## [a2.6.0 - 2022-08-12]
### Added
  - Generic Aquifer: CFI and GUI; examples illustrating applications
  - Permeability Sampler (for Seal Horizon component)
  - Thickness Sampler (for Seal Horizon component)
  - Added instructions for non-Anaconda installation
  - GUI: dynamic input through relative path
  - CFI: 3d interpolation setup for LUTR and wellbore components connected to it
  - CFI: setup of additional metric outputs for LUTR component
### Updated
  - Seal Horizon component (due to changes in the standalone Seal Flux code):
  class, CFI and GUI

source/GUI/dictionarydata.py
"""
This file consists of all the variables that are globally accessible.
Some of the variables are constants.
"""
d = {}
componentVars = {}


# componentChoices contains the names of the components already added to the simulation and
# to which connections can be created
componentChoices = []
# componentTypeDictionary contains types of the components added to the simulation in the order they were added
componentTypeDictionary = []
# connectionsDictionary contains names of the components to which the connections were created
# by the components at the given index position
connectionsDictionary = []
# connections is updated with the names of new components upon addition of the latter to the simulation
# Contains unique names of the possible connections + Dynamic Parameters option
connections = ['Dynamic Parameters']
# connectionTypes contains names of the possible connections for the components
# of the system starting with the second added component and ending with the possible
# connections for the next component (to be added yet)
connectionTypes = []

savedDictionary = {}

# Width of some widgets
# Width of labels for distribution parameters, e.g. Value:, Mean:, etc.
DISTRIBUTION_ARG_LABEL_WIDTH = 8
# Width of text fields (entries) for distribution parameters
DISTRIBUTION_ARG_TEXTFIELD_WIDTH = 8
# Width of menu containing distribution options
DISTRIBUTION_MENU_WIDTH = 16
# Width of buttons on most of the pages
BUTTON_WIDTH = 18
# Width of entries for file path input
FILE_ENTRY_WIDTH = 40
# Width of labels containing names of parameters for all components except
# generalized flow rate component and stratigraphy
PARAMETER_LABEL_WIDTH = 31
# Width of short labels containing names of component outputs
OUTPUT_LABEL_WIDTH1 = 25
# Width of long labels containing names of component outputs
OUTPUT_LABEL_WIDTH2 = 40
# Width of longer labels containing names of component outputs
OUTPUT_LABEL_WIDTH3 = 45
# Padding added on the left and right of the output checkboxes
CB_PADX = (0, 1)
# Padding added on the left and right of the parameter frames
PARAMETER_FRAME_PADX = 5
# Width of labels containing names of parameters of generalized flow rate component
GFR_PARAMETER_LABEL_WIDTH = 37
# Width of labels containing names of parameters of stratigraphy
STRATA_PARAMETER_LABEL_WIDTH = 25
# Width of labels containing names of parameters of fault leakage component
FL_PARAMETER_LABEL_WIDTH = 35
# Width of labels containing names of parameters of hydrocarbon leakage component
HCL_PARAMETER_LABEL_WIDTH = 35


# Width of widgets on Model tab
MODEL_TAB_LARGE_LABEL_WIDTH = 30
MODEL_TAB_LABEL_WIDTH1 = 25
MODEL_TAB_LABEL_WIDTH2 = 20
MODEL_TAB_LABEL_WIDTH3 = 8
MODEL_TAB_ENTRY_WIDTH = 8
MODEL_TAB_MENU_WIDTH = 10

# Width of widgets on Add Components tab
SETUP_LABEL_WIDTH = 20
SETUP_ENTRY_WIDTH = 26
SETUP_MENU_WIDTH = 20

# Setup width of widgets on the PostProcessor_Page
POSTPROCESSOR_LABEL_WIDTH = 20
POSTPROCESSOR_SPINBOX_WIDTH = 25
POSTPROCESSOR_ENTRY_WIDTH = 40
POSTPROCESSOR_FILE_ENTRY_WIDTH = 40
POSTPROCESSOR_CHECKBOX_WIDTH = 20
POSTPROCESSOR_MENU_WIDTH = 25

# Default aquifers to which leakage is possible unless number of shale layers
# is greater than 3
aquifers = ['aquifer1', 'aquifer2']

# Types of analysis available in NRAP-Open-IAM
ANALYSIS_TYPES = ["Forward", "LHS", "Parstudy"]

# Levels of logging available in NRAP-Open-IAM
LOGGING_TYPES = ["Debug", "Info", "Warning", "Error"]

# Types of distribution available for most of the parameters
# Remove Normal and lognormal for now since they cannot be bounded by lower and upper boundaries
# and there is no equivalent of truncated lognormal distribution
#DISTRIBUTION_OPTIONS = ['Fixed Value', 'Uniform', 'Normal',
#                       'Lognormal', 'Truncated', 'Triangular', 'Discrete']
DISTRIBUTION_OPTIONS = ['Fixed Value', 'Uniform', 'Truncated',
                        'Triangular', 'Discrete']

SPEC_DISTRIBUTION_OPTIONS = ['Fixed Value', 'List', 'File Input']

# Labels for possible parameters characterizing a particular distribution
DISTRIBUTION_PARS_LABELS = {
    'Fixed Value': ['Value:'],
    'List':        ['Value(s):'],
    'Uniform':     ['Min:', 'Max:'],
    'Normal':      ['Mean:', 'SD:'],
    'Lognormal':   ['Mean:', 'SD:'],
    'Truncated':   ['Mean:', 'SD:', 'Min:', 'Max:'],
    'Triangular':  ['Min:', 'Max:', 'Mode:'],
    'Discrete':    ['Values:', 'Weights:']}

# Texts for tooltips corresponding to different distribution parameters
DISTRIBUTION_PARS_SETUPS = {
    'Value:':   ['value', 'Set value of {0}.'],
    'Value(s):': ['ordvalues', 'Set list of ordered values for {0}.'],
    'File Input': ['filename', 'Select file containing values for {0}.'],
    'Min:': ['min', 'Set minimum of {0}.'],
    'Max:': ['max', 'Set maximum of {0}.'],
    'Mean:':    ['mean', 'Set mean of {0}.'],
    'SD:': ['std', 'Set standard deviation of {0}.'],
    'Mode:':    ['mode', 'Set the most common value of {0}.'],
    'Values:':  ['values', 'Set list of values for {0}.'],
    'Weights:': ['weights', 'Set weights for each value entered in Values field.']}

# Components available in the current version of GUI for NRAP-Open-IAM
COMPONENT_TYPES = [
    'Analytical Reservoir',
    'Lookup Table Reservoir',
    'Generic Reservoir',
    'Theis Reservoir',
    # 'Simple Reservoir',
    'Multisegmented Wellbore',
    'Cemented Wellbore',
    'Cemented Wellbore (WR)',
    'Open Wellbore',
    'Generalized Flow Rate',
    'Fault Flow',
    'Fault Leakage',
    'Hydrocarbon Leakage',
    'Seal Horizon',
    'Carbonate Aquifer',
    # 'Alluvium Aquifer (LF)',
    'Deep Alluvium Aquifer',
    #'Deep Alluvium Aquifer (ML)',
    'FutureGen2 Aquifer',
    'FutureGen2 AZMI',
    'Generic Aquifer',
    'Atmospheric ROM',
    'Plume Stability',
    'Chemical Well Sealing']

# Types of processing/plotting analysis and/or capabilities available for a given simulation
# Plotting is default and available for all simulations
processingTypes = ['Plotting']
APP_SIZE = [1170, 750]
TAB_SIZE = [700, 550]

# Types of plots available for a given simulation
plotTypes = ['Time Series']

# Setup of fonts for different widgets in NRAP-Open-IAM
LABEL_FONT = ('Helvetica', 14, 'bold')
LABEL_FONT2 = ('Helvetica', 12, 'bold')
INSTRUCTIONS_FONT = ('Helvetica', 12, 'bold')
DASHBOARD_FONT = ('Helvetica', 12)
